src/Grpc.AspNetCore.Server/Internal/GrpcProtocolHelpers.cs

```diff
@@ -163,5 +171,63 @@ public static void AppendStatusTrailers(HttpResponse response, Status status)
             response.AppendTrailer(GrpcProtocolConstants.StatusTrailer, status.StatusCode.ToTrailerString());
             response.AppendTrailer(GrpcProtocolConstants.MessageTrailer, status.Detail);
         }
+
+        public static AuthContext CreateAuthContext(X509Certificate2 clientCertificate)
```

Comment:
I'm going to veer a little here. If this client certificate authentication (which is coming to asp.net core via Chris and I, in which case there's functional duplication here), then it's not exactly correct. Certificates have usages (EKUs). You're not checking if the certificate is valid for use as a client authentication certificate, and that check should be performed.

This is done via checking the EKUs directly (bad) or using Validation Policies (good, because you can also check for revocation and expiry at the same time). See https://github.com/blowdart/idunno.Authentication/blob/master/src/idunno.Authentication.Certificate/CertificateAuthenticationHandler.cs#L178 for an example.

Reply:
The certificate used is comes from `HttpContext.Connection.ClientCertificate`

The C core C# gRPC implementation has an enum for setting how client certificates are handled: `new SslServerCredentials(pair, File.ReadAllText(@"Certs\ca.crt"), SslClientCertificateRequestType.RequestAndRequireAndVerify)`

It seems similar to ASP.NET Core's ClientCertificateMode: `httpsOptions.ClientCertificateMode = ClientCertificateMode.RequireCertificate;`

Reply:
So that, to my mind, is not enough. Especially when it blindly accepts any type of certificate. Hell, I could use an SMIME certificate meant for signing email. This goes against all best practice, and a bunch of SDL requirements. This needs a wider discussion outside of this PR about what you're trying to achieve here, compatibility or actual security.

Reply:
Where's the implementation for SslClientCertificateRequestType.RequestAndRequireAndVerify?

Reply:
Lets discuss offline at meeting on Monday.

Reply:
> Where's the implementation for SslClientCertificateRequestType.RequestAndRequireAndVerify?

It will be deep in some C++ code somewhere, or in another library like OpenSSL/BoringSSL. I will see if I can find it.